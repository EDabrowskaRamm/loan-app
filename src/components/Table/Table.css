.Table {
    margin: 0 auto;
    min-width: 520px;
}

thead {
    font-weight: 600;
}

tbody tr:nth-child(2n) {
    background-color: #e9e9ea;
}

td {
    padding: 5px;
}