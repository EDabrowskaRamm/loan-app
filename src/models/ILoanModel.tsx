export interface ILoanModel {
    amount: number;
    time: number;
    interest: number;
}