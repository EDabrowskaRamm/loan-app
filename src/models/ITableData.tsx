export interface ITableData {
    month: number;
    initialCapitalBalance: number;
    paymentOfInterest: number;
    creditRepayment: number;
    fullInstallment: number;
    remainingOutstanding: number;
}