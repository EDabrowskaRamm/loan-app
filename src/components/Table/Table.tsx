import React from 'react';
import './Table.css';

import { ITableData } from '../../models/ITableData'

interface IProps { data: ITableData[] }

const Table = (props: IProps) => (
    <table className="Table">
        <thead>
            <tr>
                <td>Month</td>
                <td>Initial capital balance</td>
                <td>Payment of interest</td>
                <td>Credit repayment</td>
                <td>Full installment</td>
                <td>Remaining outstanding</td>
            </tr>
        </thead>
        <tbody>
            {props.data.map(el => <tr key={el.month}>
                <td>{el.month}</td>
                <td>{el.initialCapitalBalance.toFixed(2)}</td>
                <td>{el.paymentOfInterest.toFixed(2)}</td>
                <td>{el.creditRepayment.toFixed(2)}</td>
                <td>{el.fullInstallment.toFixed(2)}</td>
                <td>{el.remainingOutstanding.toFixed(2)}</td>
            </tr>)}
        </tbody>
    </table>
);

export default Table;