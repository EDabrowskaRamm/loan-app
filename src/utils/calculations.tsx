import { ILoanModel } from '../models/ILoanModel'
import { ITableData } from '../models/ITableData'
import { InstalmentType } from '../components/App/App'

export const getRate = (loan: ILoanModel): string => {
    const getQ = 1 + ((loan.interest / 100) / 12);
    const getNumberOfMonths = loan.time * 12;
    const getRate = loan.amount * Math.pow(getQ, getNumberOfMonths) * (getQ - 1) / (Math.pow(getQ, getNumberOfMonths) - 1);
    return getRate.toFixed(2);
}

export const getTableData = (loan: ILoanModel, instalmentType: InstalmentType): ITableData[] => {
    const dataArr: ITableData[] = [];
    let data: ITableData = {
        month: 0,
        initialCapitalBalance: loan.amount,
        paymentOfInterest: 0,
        creditRepayment: 0,
        fullInstallment: 0,
        remainingOutstanding: 0
    }

    if (instalmentType === "fixed") {
        let equity: number = 0;

        for (let i = 0; i < loan.time * 12; i++) {
            const initial: number = loan.amount - equity;
            const interest: number = initial * (loan.interest / 100) * (365 / 12) / 365;
            const thisMonthsInterest: number = initial * (loan.interest / 100 / 12);

            equity += (Number(getRate(loan)) - thisMonthsInterest);

            data = {
                month: i + 1,
                initialCapitalBalance: initial,
                paymentOfInterest: interest,
                creditRepayment: (Number(getRate(loan)) - thisMonthsInterest),
                fullInstallment: (Number(getRate(loan)) - thisMonthsInterest) + interest,
                remainingOutstanding: loan.amount - equity
            }
            dataArr.push(data);
        }
    } else {
        let equity: number = loan.amount;

        for (let i = 0; i < loan.time * 12; i++) {
            const interest: number = ((loan.amount - (i * getDeclinedInstalmentEquity(loan))) * (loan.interest / 100)) / 12;
            const initial: number = equity;

            equity -= getDeclinedInstalmentEquity(loan);

            data = {
                month: i + 1,
                initialCapitalBalance: initial,
                paymentOfInterest: interest,
                creditRepayment: getDeclinedInstalmentEquity(loan),
                fullInstallment: getDeclinedInstalmentEquity(loan) + interest,
                remainingOutstanding: equity
            }
            dataArr.push(data);
        }
    }

    return dataArr;
}


const getDeclinedInstalmentEquity = (loan: ILoanModel): number => {
    return loan.amount / (loan.time * 12);
}

export const getDeclinedInstalmentInterestRate = (loan: ILoanModel): number => {
    let result: number = 0;

    for (let i = 0; i < loan.time * 12; i++) {
        result += ((loan.amount - (i * getDeclinedInstalmentEquity(loan))) * (loan.interest / 100)) / 12;
    }

    return result;
}

export const getMinMaxDeclinedInstalmentRate = (loan: ILoanModel) => {
    const data: ITableData[] = getTableData(loan, "declining");
    const firstEl = data.shift();
    const lastEl = data.pop();

    return {
        min: lastEl ? lastEl.fullInstallment : 0,
        max: firstEl ? firstEl.fullInstallment : 0
    }
}