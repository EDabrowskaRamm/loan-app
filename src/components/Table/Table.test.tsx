import React from 'react';
import renderer from 'react-test-renderer'

import Table from './Table';
import { ILoanModel } from '../../models/ILoanModel'
import { getTableData } from '../../utils/calculations'

describe('Table', () => {
    const loan: ILoanModel = {
        amount: 100000,
        time: 2,
        interest: 3.5
    }

    it('Snapshot', () => {
        const component = renderer.create(<Table data={getTableData(loan)} />)
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
});