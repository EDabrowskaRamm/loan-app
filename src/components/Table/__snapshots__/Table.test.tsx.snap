// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Snapshot 1`] = `
<table
  className="Table"
>
  <thead>
    <tr>
      <td>
        Month
      </td>
      <td>
        Initial capital balance
      </td>
      <td>
        Payment of interest
      </td>
      <td>
        Credit repayment
      </td>
      <td>
        Full installment
      </td>
      <td>
        Remaining outstanding
      </td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        1
      </td>
      <td>
        100000.00
      </td>
      <td>
        291.67
      </td>
      <td>
        4166.67
      </td>
      <td>
        4458.33
      </td>
      <td>
        95833.33
      </td>
    </tr>
    <tr>
      <td>
        2
      </td>
      <td>
        95833.33
      </td>
      <td>
        279.51
      </td>
      <td>
        4166.67
      </td>
      <td>
        4446.18
      </td>
      <td>
        91666.67
      </td>
    </tr>
    <tr>
      <td>
        3
      </td>
      <td>
        91666.67
      </td>
      <td>
        267.36
      </td>
      <td>
        4166.67
      </td>
      <td>
        4434.03
      </td>
      <td>
        87500.00
      </td>
    </tr>
    <tr>
      <td>
        4
      </td>
      <td>
        87500.00
      </td>
      <td>
        255.21
      </td>
      <td>
        4166.67
      </td>
      <td>
        4421.88
      </td>
      <td>
        83333.33
      </td>
    </tr>
    <tr>
      <td>
        5
      </td>
      <td>
        83333.33
      </td>
      <td>
        243.06
      </td>
      <td>
        4166.67
      </td>
      <td>
        4409.72
      </td>
      <td>
        79166.67
      </td>
    </tr>
    <tr>
      <td>
        6
      </td>
      <td>
        79166.67
      </td>
      <td>
        230.90
      </td>
      <td>
        4166.67
      </td>
      <td>
        4397.57
      </td>
      <td>
        75000.00
      </td>
    </tr>
    <tr>
      <td>
        7
      </td>
      <td>
        75000.00
      </td>
      <td>
        218.75
      </td>
      <td>
        4166.67
      </td>
      <td>
        4385.42
      </td>
      <td>
        70833.33
      </td>
    </tr>
    <tr>
      <td>
        8
      </td>
      <td>
        70833.33
      </td>
      <td>
        206.60
      </td>
      <td>
        4166.67
      </td>
      <td>
        4373.26
      </td>
      <td>
        66666.67
      </td>
    </tr>
    <tr>
      <td>
        9
      </td>
      <td>
        66666.67
      </td>
      <td>
        194.44
      </td>
      <td>
        4166.67
      </td>
      <td>
        4361.11
      </td>
      <td>
        62500.00
      </td>
    </tr>
    <tr>
      <td>
        10
      </td>
      <td>
        62500.00
      </td>
      <td>
        182.29
      </td>
      <td>
        4166.67
      </td>
      <td>
        4348.96
      </td>
      <td>
        58333.33
      </td>
    </tr>
    <tr>
      <td>
        11
      </td>
      <td>
        58333.33
      </td>
      <td>
        170.14
      </td>
      <td>
        4166.67
      </td>
      <td>
        4336.81
      </td>
      <td>
        54166.67
      </td>
    </tr>
    <tr>
      <td>
        12
      </td>
      <td>
        54166.67
      </td>
      <td>
        157.99
      </td>
      <td>
        4166.67
      </td>
      <td>
        4324.65
      </td>
      <td>
        50000.00
      </td>
    </tr>
    <tr>
      <td>
        13
      </td>
      <td>
        50000.00
      </td>
      <td>
        145.83
      </td>
      <td>
        4166.67
      </td>
      <td>
        4312.50
      </td>
      <td>
        45833.33
      </td>
    </tr>
    <tr>
      <td>
        14
      </td>
      <td>
        45833.33
      </td>
      <td>
        133.68
      </td>
      <td>
        4166.67
      </td>
      <td>
        4300.35
      </td>
      <td>
        41666.67
      </td>
    </tr>
    <tr>
      <td>
        15
      </td>
      <td>
        41666.67
      </td>
      <td>
        121.53
      </td>
      <td>
        4166.67
      </td>
      <td>
        4288.19
      </td>
      <td>
        37500.00
      </td>
    </tr>
    <tr>
      <td>
        16
      </td>
      <td>
        37500.00
      </td>
      <td>
        109.37
      </td>
      <td>
        4166.67
      </td>
      <td>
        4276.04
      </td>
      <td>
        33333.33
      </td>
    </tr>
    <tr>
      <td>
        17
      </td>
      <td>
        33333.33
      </td>
      <td>
        97.22
      </td>
      <td>
        4166.67
      </td>
      <td>
        4263.89
      </td>
      <td>
        29166.67
      </td>
    </tr>
    <tr>
      <td>
        18
      </td>
      <td>
        29166.67
      </td>
      <td>
        85.07
      </td>
      <td>
        4166.67
      </td>
      <td>
        4251.74
      </td>
      <td>
        25000.00
      </td>
    </tr>
    <tr>
      <td>
        19
      </td>
      <td>
        25000.00
      </td>
      <td>
        72.92
      </td>
      <td>
        4166.67
      </td>
      <td>
        4239.58
      </td>
      <td>
        20833.33
      </td>
    </tr>
    <tr>
      <td>
        20
      </td>
      <td>
        20833.33
      </td>
      <td>
        60.76
      </td>
      <td>
        4166.67
      </td>
      <td>
        4227.43
      </td>
      <td>
        16666.67
      </td>
    </tr>
    <tr>
      <td>
        21
      </td>
      <td>
        16666.67
      </td>
      <td>
        48.61
      </td>
      <td>
        4166.67
      </td>
      <td>
        4215.28
      </td>
      <td>
        12500.00
      </td>
    </tr>
    <tr>
      <td>
        22
      </td>
      <td>
        12500.00
      </td>
      <td>
        36.46
      </td>
      <td>
        4166.67
      </td>
      <td>
        4203.13
      </td>
      <td>
        8333.33
      </td>
    </tr>
    <tr>
      <td>
        23
      </td>
      <td>
        8333.33
      </td>
      <td>
        24.31
      </td>
      <td>
        4166.67
      </td>
      <td>
        4190.97
      </td>
      <td>
        4166.67
      </td>
    </tr>
    <tr>
      <td>
        24
      </td>
      <td>
        4166.67
      </td>
      <td>
        12.15
      </td>
      <td>
        4166.67
      </td>
      <td>
        4178.82
      </td>
      <td>
        -0.00
      </td>
    </tr>
  </tbody>
</table>
`;
